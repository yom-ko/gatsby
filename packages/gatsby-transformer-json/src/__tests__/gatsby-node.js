const Promise = require(`bluebird`)

const { onCreateNode } = require(`../gatsby-node`)

describe(`Process JSON nodes correctly`, () => {
  const node = {
    name: `nodeName`,
    id: `whatever`,
    parent: `SOURCE`,
    children: [],
    internal: {
      contentDigest: `whatever`,
      mediaType: `application/json`,
      name: `test`,
    },
  }

  // Make some fake functions its expecting.
  const loadNodeContent = node => Promise.resolve(node.content)

  it(`correctly creates nodes from JSON which is an array of objects`, async () => {
    const data = [
      { id: `foo`, blue: true, funny: `yup` },
      { blue: false, funny: `nope` },
    ]
    node.content = JSON.stringify(data)

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const actions = { createNode, createParentChildLink }
    const createNodeId = jest.fn()
    createNodeId.mockReturnValue(`uuid-from-gatsby`)

    await onCreateNode({
      node,
      loadNodeContent,
      actions,
      createNodeId,
    }).then(() => {
      expect(createNode.mock.calls).toMatchSnapshot()
      expect(createParentChildLink.mock.calls).toMatchSnapshot()
      expect(createNode).toHaveBeenCalledTimes(2)
      expect(createParentChildLink).toHaveBeenCalledTimes(2)
    })
  })

  it(`correctly creates a node from JSON which is a single object`, async () => {
    const data = { id: `foo`, blue: true, funny: `yup` }
    node.content = JSON.stringify(data)
    node.dir = `/tmp/foo/`

    const createNode = jest.fn()
    const createParentChildLink = jest.fn()
    const actions = { createNode, createParentChildLink }
    const createNodeId = jest.fn()
    createNodeId.mockReturnValue(`uuid-from-gatsby`)

    await onCreateNode({
      node,
      loadNodeContent,
      actions,
      createNodeId,
    }).then(() => {
      expect(createNode.mock.calls).toMatchSnapshot()
      expect(createParentChildLink.mock.calls).toMatchSnapshot()
      expect(createNode).toHaveBeenCalledTimes(1)
      expect(createParentChildLink).toHaveBeenCalledTimes(1)
    })
  })

  it(`correctly sets node type for array of objects`, async () => {
    ;[
      {
        typeName: null,
        expectedNodeTypes: [`NodeNameJson`, `NodeNameJson`],
      },
      {
        typeName: `fixed`,
        expectedNodeTypes: [`fixed`, `fixed`],
      },
      {
        typeName: (({ node, object }) => object.funny),
        expectedNodeTypes: [`yup`, `nope`],
      },
    ].forEach(async ({ typeName, expectedNodeTypes: [expectedOne, expectedTwo] }) => {
      const data = [
        { id: `foo`, blue: true, funny: `yup` },
        { blue: false, funny: `nope` },
      ]
      node.content = JSON.stringify(data)

      const createNode = jest.fn()
      const createParentChildLink = jest.fn()
      const actions = { createNode, createParentChildLink }
      const createNodeId = jest.fn()
      createNodeId.mockReturnValue(`uuid-from-gatsby`)

      await onCreateNode({
        node,
        loadNodeContent,
        actions,
        createNodeId,
      }, {
        typeName,
      }).then(() => {
        expect(createNode).toBeCalledWith(
          expect.objectContaining({
            internal: expect.objectContaining({
              type: expectedOne,
            }),
          })
        )
        expect(createNode).toBeCalledWith(
          expect.objectContaining({
            internal: expect.objectContaining({
              type: expectedTwo,
            }),
          })
        )
      })
    })
  })

  it(`correctly sets node type for single object`, async () => {
    ;[
      {
        typeName: null,
        expectedNodeType: `FooJson`,
      },
      {
        typeName: `fixed`,
        expectedNodeType: `fixed`,
      },
      {
        typeName: (({ node, object }) => object.funny),
        expectedNodeType: `yup`,
      },
    ].forEach(async ({ typeName, expectedNodeType }) => {
      const data = { id: `foo`, blue: true, funny: `yup` }
      node.content = JSON.stringify(data)
      node.dir = `/tmp/foo/`

      const createNode = jest.fn()
      const createParentChildLink = jest.fn()
      const actions = { createNode, createParentChildLink }
      const createNodeId = jest.fn()
      createNodeId.mockReturnValue(`uuid-from-gatsby`)

      await onCreateNode({
        node,
        loadNodeContent,
        actions,
        createNodeId,
      }, {
        typeName,
      }).then(() => {
        expect(createNode).toBeCalledWith(
          expect.objectContaining({
            internal: expect.objectContaining({
              type: expectedNodeType,
            }),
          })
        )
      })
    })
  })
})
