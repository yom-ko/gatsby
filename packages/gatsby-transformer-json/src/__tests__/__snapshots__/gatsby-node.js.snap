// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Process JSON nodes correctly correctly creates a node from JSON which is a single object 1`] = `
Array [
  Array [
    Object {
      "blue": true,
      "children": Array [],
      "funny": "yup",
      "id": "foo",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "FooJson",
      },
      "parent": "whatever",
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates a node from JSON which is a single object 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "blue": true,
        "children": Array [],
        "funny": "yup",
        "id": "foo",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "FooJson",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "{\\"id\\":\\"foo\\",\\"blue\\":true,\\"funny\\":\\"yup\\"}",
        "dir": "/tmp/foo/",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "File",
        },
        "name": "nodeName",
        "parent": "SOURCE",
      },
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates a node from JSON which is a single object and doesn't come from fs 1`] = `
Array [
  Array [
    Object {
      "blue": true,
      "children": Array [],
      "funny": "yup",
      "id": "foo",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "NotFileJson",
      },
      "parent": "whatever",
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates a node from JSON which is a single object and doesn't come from fs 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "blue": true,
        "children": Array [],
        "funny": "yup",
        "id": "foo",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "NotFileJson",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "{\\"id\\":\\"foo\\",\\"blue\\":true,\\"funny\\":\\"yup\\"}",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "NotFile",
        },
        "parent": "SOURCE",
      },
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates nodes from JSON which is an array of objects 1`] = `
Array [
  Array [
    Object {
      "blue": true,
      "children": Array [],
      "funny": "yup",
      "id": "foo",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "NodeNameJson",
      },
      "parent": "whatever",
    },
  ],
  Array [
    Object {
      "blue": false,
      "children": Array [],
      "funny": "nope",
      "id": "uuid-from-gatsby",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "NodeNameJson",
      },
      "parent": "whatever",
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates nodes from JSON which is an array of objects 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "blue": true,
        "children": Array [],
        "funny": "yup",
        "id": "foo",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "NodeNameJson",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "[{\\"id\\":\\"foo\\",\\"blue\\":true,\\"funny\\":\\"yup\\"},{\\"blue\\":false,\\"funny\\":\\"nope\\"}]",
        "dir": "/tmp/foo/",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "File",
        },
        "name": "nodeName",
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "blue": false,
        "children": Array [],
        "funny": "nope",
        "id": "uuid-from-gatsby",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "NodeNameJson",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "[{\\"id\\":\\"foo\\",\\"blue\\":true,\\"funny\\":\\"yup\\"},{\\"blue\\":false,\\"funny\\":\\"nope\\"}]",
        "dir": "/tmp/foo/",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "File",
        },
        "name": "nodeName",
        "parent": "SOURCE",
      },
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates nodes from JSON which is an array of objects and doesn't come from fs 1`] = `
Array [
  Array [
    Object {
      "blue": true,
      "children": Array [],
      "funny": "yup",
      "id": "foo",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "NotFileJson",
      },
      "parent": "whatever",
    },
  ],
  Array [
    Object {
      "blue": false,
      "children": Array [],
      "funny": "nope",
      "id": "uuid-from-gatsby",
      "internal": Object {
        "contentDigest": "contentDigest",
        "type": "NotFileJson",
      },
      "parent": "whatever",
    },
  ],
]
`;

exports[`Process JSON nodes correctly correctly creates nodes from JSON which is an array of objects and doesn't come from fs 2`] = `
Array [
  Array [
    Object {
      "child": Object {
        "blue": true,
        "children": Array [],
        "funny": "yup",
        "id": "foo",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "NotFileJson",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "[{\\"id\\":\\"foo\\",\\"blue\\":true,\\"funny\\":\\"yup\\"},{\\"blue\\":false,\\"funny\\":\\"nope\\"}]",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "NotFile",
        },
        "parent": "SOURCE",
      },
    },
  ],
  Array [
    Object {
      "child": Object {
        "blue": false,
        "children": Array [],
        "funny": "nope",
        "id": "uuid-from-gatsby",
        "internal": Object {
          "contentDigest": "contentDigest",
          "type": "NotFileJson",
        },
        "parent": "whatever",
      },
      "parent": Object {
        "children": Array [],
        "content": "[{\\"id\\":\\"foo\\",\\"blue\\":true,\\"funny\\":\\"yup\\"},{\\"blue\\":false,\\"funny\\":\\"nope\\"}]",
        "id": "whatever",
        "internal": Object {
          "contentDigest": "whatever",
          "mediaType": "application/json",
          "type": "NotFile",
        },
        "parent": "SOURCE",
      },
    },
  ],
]
`;
